core-rust/state-manager/src/mempool/mempool_relay_dispatcher.rs

```diff
@@ -0,0 +1,116 @@
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use jni::errors::Result;
+use jni::objects::{GlobalRef, JObject, JValue};
+use jni::{JNIEnv, JavaVM};
+use radix_engine_common::data::scrypto::scrypto_encode;
+use std::ops::Deref;
+
+use transaction::model::NotarizedTransaction;
+
+use crate::jni::mempool::JavaRawTransaction;
+use crate::jni::utils::jni_slice_to_jbytearray;
+
+/// A Java dispatcher for a "new transaction from Core API" event.
+pub struct MempoolRelayDispatcher {
+    jvm: JavaVM,
+    j_state_manager_ref: GlobalRef,
+}
+
+impl MempoolRelayDispatcher {
+    /// Creates a long-lived dispatcher from the given short-lived JNI context and Java state
+    /// manager reference.
+    pub fn new(env: &JNIEnv, j_state_manager: JObject) -> Result<Self> {
+        Ok(Self {
+            jvm: env.get_java_vm()?,
+            j_state_manager_ref: env.new_global_ref(j_state_manager)?,
+        })
+    }
+
+    /// Triggers a relay of the given transaction.
+    /// This is implemented using an event dispatch on the Java side, so this method should not
+    /// block. Any Java exceptions will be returned as `Err` (i.e. their Java exception status will
+    /// be cleared, as if this method was catching them).
+    pub fn trigger_relay(&self, transaction: NotarizedTransaction) -> Result<()> {
+        let attachment = self.jvm.attach_current_thread()?;
```

Comment:
Do we need to detach after the call? or I guess this handled as part of Rust drop of attachment?